### ISSUE TITLE
Proton Calendar Integration

### ISSUE THREAD

**Issue by xdaluca** [calcom/cal.com#5756]
### Is your proposal related to a problem?

for [Proton Calendar](https://proton.me/calendar) users, need to use .ics

### Describe the solution you'd like

integration with [proton calendar](https://proton.me/calendar) for availabilities and meeting confirmations, finally I can give up on google altogether.

### Describe alternatives you've considered

manual process, or Google.

### Additional context

the topic has been raised at https://github.com/calcom/cal.com/issues/3167

**Comment by PeerRich** [comment]
@alishaz-polymath can you check if proton runs on CalDav?

**Comment by PeerRich** [comment]
maybe also sync proton with a different calendar that is compatible in the meantime?

**Comment by alishaz-polymath** [comment]
Sure. I'll take a look. 👍

**Comment by shawnyeager** [comment]
FYI that Proton now provides the ability to create a CalDav (.ics) URL, however, it does not need/require a username and password, which the Cal.com CalDav app requires be populated.

**Comment by alishaz-polymath** [comment]
Hey @shawnyeager 
I had a look at Proton and I could see that the .ics is **read-only** which suggests, cal events will not be created on the calendar using the .ics link that you mentioned. Can you confirm if this is the case or have I misunderstood it?
ref: https://proton.me/support/share-calendar

**Comment by shawnyeager** [comment]
Thanks, @alishaz-polymath. That's correct. Their design and encryption prevents 3rd-party writes currently. 

I was hoping to at least be able to read for inclusion in availability.

**Comment by alishaz-polymath** [comment]
> Thanks, @alishaz-polymath. That's correct. Their design and encryption prevents 3rd-party writes currently.
> 
> I was hoping to at least be able to read for inclusion in availability.

I have contacted them to see their stance on our potential integration. I"ll wait to hear back from them and proceed accordingly. The fact that it can only receive freebusy data suggests potential issues later on since this cannot be used as a destination calendar. I'll look further into it and see if we can make it work for just freebusy check. 🙏 
I'll keep you guys posted here

**Comment by alishaz-polymath** [comment]
So, I've had a chat for their Tech Lead and we are currently trying to figure out the mechanics of providing a proper integration, not just the fetch of freebusy data, but a proper two-way sync.
They should get back to me with further details at some point this week (I think) and I would be able to provide more information once they do 🙏

**Comment by shawnyeager** [comment]
Great news. You'll be the only product that I'm aware of, so among the privacy-focused crowd, you'll have a strong differentiation. Thanks.

**Comment by xdaluca** [comment]
> Great news. You'll be the only product that I'm aware of, so among the privacy-focused crowd, you'll have a strong differentiation. Thanks.

Indeed, also I think that given the moves that the Proton team have been making with adding a calendar and drive to the email, maybe shared docs in the future, adoption is rising.

**Comment by PeerRich** [comment]
excited for this. great stuff @alishaz-polymath

**Comment by PeerRich** [comment]
if anyone wants to pick this up, here's how to make an app: https://docs.cal.com/how-to-guides/how-to-build-an-app

**Comment by aar2dee2** [comment]
Hey, could you assign this to me?

**Comment by alishaz-polymath** [comment]
Hi @aar2dee2 we're currently in talks with the team at Proton Calendar as there's no direct way of doing it just yet. As there needs to be internal discussions and talks between Cal and Proton, potentially, I think it's better if we handle it atleast until we have a plan in place 🙏

**Comment by aar2dee2** [comment]
Sure, thanks!

**Comment by THESHade** [comment]
Please keep us posted; otherwise, it'd be great to have at least the availability. I'd even try to look into it (found this thread while doing my research :) )

**Comment by alishaz-polymath** [comment]
Hi @THESHade There is development in works at Proton's end, so it's pretty positive right now and moving ahead. There was an unexpected delay on their end, but I heard back from them this week that there's potentially going to be news in a week or two 🎉 
Once that's clear, I'll make sure to update here the state of this plan. As for simply availability checking, it doesn't seem wise to invest time into a partially usable integration when development for end to end sync is currently underway

**Comment by JuliusGudni** [comment]
@alishaz-polymath Any update on this?

**Comment by alishaz-polymath** [comment]
Hey @JuliusGudni 
I have tried to get in touch with them,however, I haven't heard back yet. I'll try to chase them again just in case my email was skipped by mistake. 🙏

**Comment by Arrendy** [comment]
Excited to hear that this is in the works! Is there anything that the community can do to help move this forward? Or just need patience?

**Comment by alishaz-polymath** [comment]
Hey @Arrendy 
We just need to be patient as far as this is concerned, I'm afraid. I have contacted them again and I did hear back from their support team, they say they'll get back to me ASAP, but not really an exact ETA just yet. 🙏 
Thank you for your patience on this one, I know it has been a while, but it's not up to us at the moment, we're trying our best to push it as best as we can. 😄

**Comment by Arrendy** [comment]
Thanks for the update. There's also a feature request for Proton Calendar integrations where some cal.com love can be shown!
https://protonmail.uservoice.com/forums/932842-proton-calendar/suggestions/45821284-integration-with-other-apps-and-platforms

**Comment by alishaz-polymath** [comment]
So I just heard back from them after quite a lot of back-to-back email bumps, and unfortunately, they responded with not too promising information.

They were attempting to experiment with Caldav integration, which prompted them to provide a test API and move forward. However, as a general policy, they do not support integrations between their Proton products and other third-party services. This makes it challenging to prioritize such projects against other features on their roadmap for the time being.

There are multiple reasons for this stance, such as maintaining a consistent user experience and functionality of their services, ensuring end-to-end encryption wherever possible, preventing unexpected bugs, and guaranteeing the validity of their third-party security audits. The latter becomes particularly challenging when integrating their services with other products. For these reasons, they don't support integrations with any third-party products. It is important to understand their position that this is a uniform rule they uphold for all third parties. However, if there are any changes to this policy in the future, they will make sure to inform accordingly.

They tried, but couldn't prioritize adding CalDAV on their system (or a direct 3rd party integration with us). If you guys can make some noise on their end and be vocal to them, maybe this changes.

Closing this ticket for the time being (until their stance changes, if it does, when it does)

**Comment by alishaz-polymath** [comment]
We really wanted to give you an end-to-end experience with Proton Calendar and really tried to push them to make it work (for many months), but clearly it is not possible at the moment and personally, I'm gutted. 

So, just as a point of note, to support ics feed for availability we will be looking at issue #3167 which should help with availability integration with proton calendar at the very least.

**Comment by christopherpickering** [comment]
Thanks for trying! Maybe one day 👍🏼

**Comment by shawnyeager** [comment]
Thanks, all, for trying to make this happen.  As a long-time Proton user and subscriber, this is disappointing. I appreciate the complexities on their end, but it will be increasingly difficult for Proton to remain viable as a walled garden.

**Comment by colans** [comment]
Crazy idea:  How about just sending the usual iCal email message to all parties, but *not* writing to the calendar?  Proton users can then access their account, look at the message, and then hit the "Add to calendar" button.  It's an extra step, but makes the system usable without requiring write access.

Availability will already be handled by #3167 so no need to worry about that here.

All that would be required in user accounts is an option to not write to any calendar, which seems like a minimal amount of work.

What do you think?

**Comment by BenLewies** [comment]
@colans I support that idea.

**Comment by colans** [comment]
As we're not able to reopen this issue, I've created https://github.com/calcom/cal.com/issues/14300 as a follow-up.  Go give it a :+1: .

**Comment by PeerRich** [comment]
reopening to improve our Proton Calendar integration.

Ideally we have Proton Calendar as a standalone app in app.cal.com/apps/proton

**Comment by kesh-007** [comment]
Is the issue still open? I am interested in working on this

**Comment by DeepaPrasanna** [comment]
> app.cal.com/apps/proton

The page doesn't exists

**Comment by mogery** [comment]
> > app.cal.com/apps/proton
> 
> 
> 
> The page doesn't exists

Yes. This is an issue to create that integration. It is not completed yet.

**Comment by warengonzaga** [comment]
I shared my idea here: https://github.com/calcom/cal.com/issues/14300#issuecomment-2167174611

**Comment by warengonzaga** [comment]
> > > app.cal.com/apps/proton
> > 
> > 
> > The page doesn't exists
> 
> Yes. This is an issue to create that integration. It is not completed yet.

Would love to try it once completed!

**Comment by warengonzaga** [comment]
Merged the latest update: https://github.com/calcom/cal.com/pull/15454

**Comment by colans** [comment]
I'm still seeing the problem I mentioned at https://github.com/calcom/cal.com/issues/14300#issuecomment-2165669805 on Prod even though that issue is closed.

Would someone kindly provide an update on the status of that?  Does it require a new issue?  What has to happen for this to get fixed?  Is it on the roadmap at all?  Thanks!

**Comment by ibnunowshad** [comment]
Proton Calendar recently added features to include others .ics calendar into Proton in iCal format.

Cal.com is something I highly use day to day. In proton calendar, there is option to setup or create meeting with aliases. I have personally tested to create meeting with aliases in my SimpleLogin.

I am mot developer. However, I use different aliases various apps in the internet. So while integrating Proton calendar into Cal.com kindly consider to use the registered email address aka aliases in Cal.com :)

**Comment by pratyushkumar06** [comment]
Hi, Is this still Open? I could try and attempt this

**Comment by monty241** [comment]
Is there a hard-fork from cal.com by a company not under US law (Cal.com is under US law since located in SF)? 

Would love to add a bounty for read-only Proton Cal integration, but due to internal choices we are phasing out all software products under US control and not implementing a new one.  That is the reason we are phasing out Calendly.

**Comment by monty241** [comment]
> Hey so does somebody know if this works now?

Yes, it works. Choose ICS Feed.

**Comment by otra-ch** [comment]
> > Hey so does somebody know if this works now?
> 
> Yes, it works. Choose ICS Feed.

sure about that? I tried with ICS but cal.com doesn't have the rights to write on the calendar.

**Comment by monty241** [comment]
Yes. Just someone scheduled an appointment with me. Maybe it does not write through the ICS but through an email, but without manual actions the appointment appeared directly in my Proton Calendar:

![Image](https://github.com/user-attachments/assets/33cd709c-1125-4661-afd8-82ac72f99da7)

It was not necessary to create some automatic handling.

Please note that I have used a separate ICS calendar link for a separate calendar which only contains appointments from 2025 and further. 

The link for my calendar did not work, since the URL download crashes after 5 minutes. This has been reported to Proton Support. The normal calendar has 15.000 calendar items, the newly created 2025 calendar only few hundred.

**Comment by colans** [comment]
@monty241 wrote:
> Yes, it works. Choose ICS Feed.

No, it doesn't.  See [my comment in the other issue](https://github.com/calcom/cal.com/issues/14300#issuecomment-2165669805).

Sure, you can set it up now without writing to the calendar, but the problem with ghost events still needs to be fixed.  I think that's all that's left here.

**Comment by monty241** [comment]
@colans OK. I don't know the policy used to determine whether a feature is complete for this project and whether it must have "Proton logo" on it, but maybe it is better to:

- close this "feat" issue when the feature is there, but has a bug, 
- open a bug report for the ghost events (since #14300 is closed),
- link to the bug report from this one.

In my opinion, features would never get shipped since there is always some bug.

**Comment by monty241** [comment]
With CalDAV it also required user name and password (which are probably somewhere encoded already present in the URL) and presenting them nonetheless triggers an error.

For me it works using the ICS option (Dutch sample):

![Image](https://github.com/user-attachments/assets/3c6ef9f0-6687-4ce7-a6f9-797b297f2c77)

**Comment by colans** [comment]
@shawnyeager : When you "Add ICS Feed", and you see the pop-up, make sure to check the box in the middle:

![Image](https://github.com/user-attachments/assets/be0bbac5-c5ca-4e16-b58f-eba83b276724)

**Comment by shawnyeager** [comment]
Thanks, @colans. I'd mistakenly chosen CalDav, instead of ICS. User error.

**Comment by colans** [comment]
@monty241 wrote:
> Is there a hard-fork from cal.com by a company not under US law (Cal.com is under US law since located in SF)?

This is off-topic, but you're welcome to [take a look](https://github.com/calcom/cal.com/forks).  If you can't find one, [the code is open source](https://github.com/calcom/cal.com) so feel free to just run it yourself.

**Comment by colans** [comment]
@monty241 wrote:
> [@colans](https://github.com/colans) OK. I don't know the policy used to determine whether a feature is complete for this project and whether it must have "Proton logo" on it, but maybe it is better to:
> 
>  * close this "feat" issue when the feature is there, but has a bug,
> 
>  * open a bug report for the ghost events (since [[CAL-3738] Feat: Fix Proton Calendar with ICS #14300](https://github.com/calcom/cal.com/issues/14300) is closed),
> 
>  * link to the bug report from this one.
> 
> 
> In my opinion, features would never get shipped since there is always some bug.

I agree, but I'm wondering if the maintainers will take this more seriously if we keep the main issue open.  Let's see what they say.

**Comment by PeerRich** [comment]
would it be worth adding a proton mail app that is simply the reskinned ics app?

**Comment by warengonzaga** [comment]
> would it be worth adding a proton mail app that is simply the reskinned ics app?

That's a good idea; however, it would be a duplicate that creates a small difference.

**Comment by warengonzaga** [comment]
I added and use the ICS for proton mail however I'm getting this error: Could not book the meeting.
createEvent called on read-only ICS feed

**Comment by monty241** [comment]
> I added and use the ICS for proton mail however I'm getting this error: Could not book the meeting. createEvent called on read-only ICS feed

My co-worker reported this issue. They are working on it was the statement. Best is to report it under a separate issue. My co-worker 
didn't raise one yet.

**Comment by elioqoshi** [comment]
My workaround I use is connect Apple Calendar to ProtonCalendar via ICS and then connect Cal.com via the Apple Calendar integration. That works pretty well for Apple users, anyone else tried that?

**Comment by warengonzaga** [comment]
> > I added and use the ICS for proton mail however I'm getting this error: Could not book the meeting. createEvent called on read-only ICS feed
> 
> My co-worker reported this issue. They are working on it was the statement. Best is to report it under a separate issue. My co-worker didn't raise one yet.

After reconfiguring it works well.

**Comment by alishaz-polymath** [comment]
Hey guys, can you try the ICS integration again and confirm if the issue persists? We've made a couple of changes to the ICS feed integration and it should work as expected now 🙏

**Comment by colans** [comment]
@alishaz-polymath : Thanks for looking into this!  No checkbox required for writes now, but I'm still getting the ghost events though, which block bookings for the first several days (originally reported [here](https://github.com/calcom/cal.com/issues/14300#issuecomment-2165669805)) :

![Image](https://github.com/user-attachments/assets/edf8f552-9888-4244-8cb8-5590dce75bbd)

**Comment by alishaz-polymath** [comment]
> > Hey guys, can you try the ICS integration again and confirm if the issue persists? We've made a couple of changes to the ICS feed integration and it should work as expected now 🙏
> 
> Is any action required before changes are visible?
> 
> It is for example still not possible to plan on the 21st of April in the calendar, eventho it is available to select:
> 
> ![Image](https://github.com/user-attachments/assets/75790354-356e-4418-9c87-e9c97607823f)
> 
> ![Image](https://github.com/user-attachments/assets/5155cd6c-90d6-445a-8079-3deeb09a1f04)

That is an unrelated error.

**Comment by alishaz-polymath** [comment]
> [@alishaz-polymath](https://github.com/alishaz-polymath) : Thanks for looking into this! No checkbox required for writes now, but I'm still getting the ghost events though, which block bookings for the first several days (originally reported [here](https://github.com/calcom/cal.com/issues/14300#issuecomment-2165669805)) :
> 
> ![Image](https://github.com/user-attachments/assets/edf8f552-9888-4244-8cb8-5590dce75bbd)

That is interesting, would you mind sharing with me a proton ICS link please? Feel free to send over the email if it is private, or create a sample public ICS for testing purposes and share 🙏 .

**Comment by colans** [comment]
@alishaz-polymath : These are just free/busy URLs so nothing private:
* https://calendar.proton.me/api/calendar/v1/url/AJ7I6a1KAd_5lMKjOKFfs_swx86vg4PtzMswOOS92fv28J2dX77Lrzyf21fI7hswwUcDP8Ef0yJLG9Az-w8xwA==/calendar.ics?CacheKey=j7L3WJO4Ht3c2uuGtcAPdg%3D%3D
* https://calendar.proton.me/api/calendar/v1/url/-xGd6BRX3FEVc9PE63g-X-PFn2Qpobs9gomj1wA4TknJgO1OhBqFYL3LfWsk9ZD3kCh2pFvt0q6HtQvRNBtrQw==/calendar.ics?CacheKey=aTVaf3p2W2ShajkIGbB-MQ%3D%3D

**Comment by Gefionious** [comment]
I am seeing a different symptom but it might be related. When I configure proton calendar using a free/busy ICS url 
`https://calendar.proton.me/api/calendar/v1/url/JG3JtRQEE2FbYUGYSh5SVnfIainLYup-n2BE7oKzxTcDzLSMtBcKKCxTi8ZfoQbPXXkstrA7zEwff-3muIA4uw==/calendar.ics?CacheKey=0ZkxEIZVARXDsxmVlmftSw%3D%3D`
In this specific case I have a meeting this Friday 4/25 8:30am-9:30am EST (NY) but it is not showing up as busy in cal.com and it allows overbooking.

**Comment by alishaz-polymath** [comment]
I will be diving further into this soon, and we're planning on adding a dedicated Proton App soon, where ideally we could take care of these nuances. No ETA yet, but it's lined up 🙏

**Comment by Gefionious** [comment]
> I will be diving further into this soon, and we're planning on adding a dedicated Proton App soon, where ideally we could take care of these nuances. No ETA yet, but it's lined up 🙏

That sounds fantastic. Please let me know if I can help in any way. I volunteer to test anytime, want to contribute if I can. Rusty on my dev skills, old C programmer from the 80's learning Python, but I can still test, lol.

**Comment by alishaz-polymath** [comment]
I will reach out for sure when it's time. Thank you, appreciate it 🙏 

> > I will be diving further into this soon, and we're planning on adding a dedicated Proton App soon, where ideally we could take care of these nuances. No ETA yet, but it's lined up 🙏
> 
> That sounds fantastic. Please let me know if I can help in any way. I volunteer to test anytime, want to contribute if I can. Rusty on my dev skills, old C programmer from the 80's learning Python, but I can still test, lol.

**Comment by itshaithamn** [comment]
Is this bounty still open? It's a long thread with multiple different issues?

**Comment by 5ajaki** [comment]
@alishaz-polymath - what’s the best approach on this?  Wait to test something or do it manually?

**Comment by alishaz-polymath** [comment]
> Is this bounty still open? It's a long thread with multiple different issues?

Which bounty are you referring to? The issue is definitely still open. The idea is to take our base ICS app and create a PROTON specific app where we handle the proton nuances, so that they don't affect ICS app (and thus avoid breaking ICS app for other providers and general ICS feeds).

I haven't been able to find enough time to tackle it yet, so if you have time feel free to give it a go! 🙌

**Comment by jayantpranjal0** [comment]
@alishaz-polymath any updates?

**Comment by op021** [comment]
/claim #5756
I'm interested in working on this issue and have already started exploring possible solutions.

**Comment by AmaanBilwar** [comment]
is this issue still open to work on?

**Comment by fmorency** [comment]
I'm getting a

```
500: INTERNAL_SERVER_ERROR Code: MIDDLEWARE_INVOCATION_FAILED
```

when trying to add the Limited View (Free/Busy) Proton calendar ICS feed.

**Comment by Piyushrathoree** [comment]
is this issue still opened ?

**Comment by MMesch** [comment]
I tried to set Proton/Cal.com integration up via ICS feed but it doesn't work: slots that are clearly taken in my proton calendar are still available in cal.com. I *think* all settings are correct but I also couldn't find clear instructions. I am also unsure how to debug this and obtain more information what the problem could be since no error messages are shown anywhere and everything seems to be connected.

**Comment by simiondolha** [comment]
Added a feat(app-store): add generic CalDAV integration with read-only support
Generic CalDAV app supporting Proton, Fastmail, Nextcloud, Synology, Zoho,
and any CalDAV server. Automatically detects read-only calendars (like Proton
with encryption) and falls back to ICS email for event confirmations.

**Comment by gasparegas** [comment]
I noticed there is already a generic CalDAV integration with read-only support.
Before starting, I’d like to clarify:

- Is the goal of this issue to provide a **native Proton Calendar integration**
  (OAuth + Proton API, read/write)?
- Or should this bounty focus on **extending the existing CalDAV app**
  (e.g. write support, better UX, Proton-specific flows)?

Once confirmed, I’ll proceed with the implementation.

**Comment by tba-gh** [comment]
I would suggest whatever method is easier to accomplish. I think what we are all looking for is a reliable read/write support.  Dealers choice on how to get there.

**Comment by gasparegas** [comment]
Perfect, thanks for confirming 👍  

I’ll proceed by extending the existing CalDAV integration to support **reliable read/write** for Proton Calendar, focusing on correctness and UX.  

I’ll start with a proposal and then open a PR for review.

**Comment by jacanidjar** [comment]
@xdaluca This bounty is available, can I start it?

**Comment by Manideepchopperla** [comment]
Hi @PeerRich, I see @gasparegas has started working on this. 
If that PR does not get merged, I’d be happy to work on this bounty. 
Please let me know if it’s still open for additional submissions.

**Comment by Rhan2020** [comment]
/attempt #5756

Opened PR #27863 — adds a dedicated Proton Calendar app based on the ICS feed app, as suggested by @PeerRich and @alishaz-polymath.

Key fixes:
- Ghost events: filters `STATUS:CANCELLED` events that Proton includes in ICS feeds
- Cancelled recurring occurrences properly skipped
- Proton URL validation (only proton.me/protonmail.com accepted)
- Proton-specific branding and setup flow

**Comment by anikdhabal** [comment]
There is no official API documentation or public API provided by Proton. Also, we need to add this integration using proper OAuth. So closing it for now 🙏

**Comment by colans** [comment]
@anikdhabal : There doesn't need to be any Proton API work because (as we've discussed earlier), users just need the iCal file (which gets sent to them) so that they can add to their calendars themselves.  We now have the option to *not* to write to calendar so let's take advantage of that.  So your reason for closing this issue isn't actually valid.

@Rhan2020 's attempt is fantastic, because it's the first I've seen that actually deals with the ghost-events problem (which is currently preventing me from using `cal.com`).  Let's **please reopen this issue**, and **take a closer look at this attempt**.  What do you think?

Keeping this closed would be unfortunate, because then we'd have to open a new issue, and start this process all over again.  This is already the second time we're doing this; let's not do it again.

**Comment by guidol-ive** [comment]
Given long thread, including direct link to former discussion that there is no and will probably never be another API different from ICS:

https://github.com/calcom/cal.com/issues/5756#issuecomment-1600571667

**Comment by realmpastai-web** [comment]
I'd like to work on this bounty

**Comment by kacperkostrzab** [comment]
Hi, I'm a TypeScript Developer with a strong background in building complex integrations and web automation. I’ve been following Cal.com for a while and I’d love to contribute by building the Proton Calendar Integration.

My experience in developing scheduling-related tools and handling API authentication (OAuth/API Keys) makes me a great fit for this task. I understand the importance of end-to-end security when dealing with Proton's ecosystem and will ensure the integration is robust, performant, and follows Cal.com's architectural patterns.

I'm available to start immediately and can provide a clear PR within a few days. Could you please assign this to me?

**Comment by twcurrie** [comment]
^ 🤖 ^

**Comment by chengyixu** [comment]
## Proposal

**Problem:** Cal.com lacks a Proton Calendar integration. Proton keeps ghost events (STATUS:CANCELLED) in ICS feeds and the generic ICS app does not handle this.

**Solution:** Build a dedicated Proton Calendar app that fetches/parses ICS data, filters ghost events, handles EXDATE for recurring events, enforces domain validation (proton.me/protonmail.com), and encrypts feed URLs at rest.

**Implementation:**
- New app: `packages/app-store/protoncalendar/`
- CalendarService with Proton-specific ICS parsing
- Setup page with URL validation
- API route for credential management

JS/TS/React developer with 30+ published npm tools. Happy to implement this.

GitHub: @chengyixu

**Comment by colans** [comment]
@anikdhabal : Any particular reason you didn't reopen [!27863](https://github.com/calcom/cal.com/pull/27863) after the discussion we had at https://github.com/calcom/cal.com/issues/5756#issuecomment-3955934100 ?

It sounds like we have something to iterate on there, and it would be better for the community to consolidate efforts instead of constantly adding more proposals.  What do you think?

**Comment by Coincatcher** [comment]
I'd like to work on this bounty. I have experience with TypeScript calendar integrations and will implement the Proton Calendar integration following the existing patterns (Google Calendar, Outlook). Expected timeline: 3-4 days.

**Comment by Coincatcher** [comment]
## Proposal: Proton Calendar Integration

Hey! I'd like to take on this bounty. I've analyzed the Cal.com codebase, the existing calendar integrations, and Proton Calendar's current technical capabilities. Here's my plan.

### Technical Analysis

**Core challenge:** Proton Calendar has **no public API** and **no CalDAV support**. Due to its end-to-end encryption architecture, Proton deliberately prevents third-party write access to calendars. This means a full read/write integration (like Google Calendar or Office 365) is architecturally impossible today.

**What IS available:**
1. **ICS share links** (paid Proton plans only) — Proton generates URLs that expose a read-only ICS feed. Two modes: "Limited view" (free/busy only) and "Full view" (all event details, requires Proton to temporarily decrypt).
2. **ICS email invitations** — PR #15454 (merged Jul 2024, issue #14300) already added a "skip writing to calendar" option to the `ics-feed` calendar app, specifically to support Proton Calendar users who want availability checking without write access.

**Current gap:** While the ICS feed workaround exists, there is no first-class "Proton Calendar" app in the app store. Users have to know about the ICS feed workaround, manually obtain their Proton share link, and configure it through the generic ICS Feed app. This is confusing and undiscoverable.

### Proposed Solution

Create a dedicated **`protoncalendar`** app-store package that wraps the existing ICS feed infrastructure with a Proton-branded, guided setup experience.

#### Implementation:

**1. Create `packages/app-store/protoncalendar/` with standard structure:**
- `api/add.ts` — Credential storage handler
- `lib/CalendarService.ts` — Extends/wraps ICSFeedCalendarService
- `pages/setup/index.tsx` — Guided setup UI with Proton-specific instructions
- `static/icon.svg` + `icon-dark.svg` — Proton Calendar branding
- `_metadata.ts`, `config.json`, `index.ts`, `DESCRIPTION.md`, `package.json`

**2. CalendarService** — Delegate to existing `ICSFeedCalendarService`, passing the Proton ICS URL as credential. Reuses all existing logic: ICS fetching, `ical.js` parsing, busy time extraction, recurring event expansion, timezone handling, and graceful no-op for write operations.

**3. Setup page** — Step-by-step onboarding:
- Instructions with screenshots showing how to generate an ICS share link in Proton Calendar
- Note that a paid Proton plan is required
- Input field for the ICS URL
- Pre-checked "Skip writing events" checkbox (locked, since Proton is read-only)
- Sync delay warning (up to 8 hours)

**4. Graceful event creation handling:**
- When Proton is the only connected calendar, bookings succeed but skip calendar event creation
- Standard `.ics` email attachments sent so users can manually import (leveraging PR #15454)
- Clear UI notice about the read-only limitation

### Files to modify/create

| Action | Path | Purpose |
|--------|------|--------|
| **Create** | `packages/app-store/protoncalendar/*` | Entire new app package (8-10 files) |
| **Modify** | `packages/app-store/apps.metadata.generated.ts` | Register app metadata |
| **Modify** | `packages/app-store/apps.generated.ts` | Register app exports |
| **Modify** | `apps/web/public/static/locales/en/common.json` | Add i18n strings |

### Estimated timeline

- **Day 1-2:** Scaffold app package, metadata, config, static assets. Setup page UI.
- **Day 3-4:** CalendarService implementation, credential handling, integration testing.
- **Day 5:** Edge cases, email .ics fallback verification, i18n, PR cleanup.

### Known limitations (transparent)

1. **Read-only** — Proton has no public API/CalDAV. Events cannot be written programmatically.
2. **Paid Proton plan required** — Calendar sharing via link requires Proton Mail Plus or higher.
3. **Sync delay up to 8h** — Proton's ICS feed is not real-time.
4. **No two-way sync** — Architecture designed for easy upgrade if Proton ever releases an API.

### Why this approach works

Rather than waiting for a Proton API that may never come, this gives Proton users a **first-class, discoverable experience** in the Cal.com app store. It reuses proven infrastructure (ICS feed service) while providing Proton-specific onboarding. The 33+ upvotes and 100+ comments show significant demand.

I'm ready to start immediately. Happy to discuss any aspect of this approach.

**Comment by SBALAVIGNESH123** [comment]
is the bounty avaiable?

**Comment by colans** [comment]
For folks like me that have been waiting for this to get prioritized (the community keeps offering solutions, and Cal has never seriously looked at them), Proton just released their own booking-page support:

* [How to use appointment scheduling in Proton Calendar](https://proton.me/support/calendar-appointment-scheduling)

So Cal.com accounts are no longer necessary for the Proton crowd, which is fine, because they weren't terribly interested in supporting us anyway.

**Comment by nehemiyawicks** [comment]
/attempt #5756

Hey, I'd like to pick this up. I've gone through all 6 prior PRs and the full
issue thread to understand what went wrong and what you actually want.

## Approach

Following alishaz-polymath's guidance (Jun 2025): "take our base ICS app and
create a PROTON specific app where we handle the proton nuances." Read-only
via Proton's ICS share link. That's the only integration path Proton supports
since they officially refused third-party API access back in June 2023.

## Plan

New directory: `packages/app-store/protoncalendar/`

Modeled on `ics-feedcalendar`, implementing the `Calendar` interface directly.

1. **`_metadata.ts`** - `type: "proton_calendar"`, `slug: "proton-calendar"`,
   `isOAuth: false`, Proton branding.

2. **`api/add.ts`** - Accept Proton ICS share URLs. Validate:
   - HTTPS only
   - Strict hostname match against `calendar.proton.me` (not `endsWith`,
     PR #28372's cubic-dev-ai flagged `evilproton.me` bypass)
   - Encrypt URLs at rest with `CALENDSO_ENCRYPTION_KEY` since the URL
     contains the CacheKey which is basically a bearer token
   - Validate by fetching + parsing before saving (same as ics-feedcalendar)

3. **`lib/CalendarService.ts`** - Implements `Calendar`:
   - `createEvent` / `updateEvent` / `deleteEvent` are no-ops with
     `calWarnings: ["Proton Calendar is read-only"]`
   - `listCalendars()` returns with `readOnly: true`
   - `getAvailability()` is the core logic:
     - Fetch ICS with `AbortSignal.timeout(10_000)` + `redirect: 'manual'`
       for SSRF mitigation (from PR #28414)
     - Parse with ical.js
     - Filter ghost events: skip `STATUS:CANCELLED`, track cancelled
       recurring instances by `uid:recurrence-id` composite key, not just
       timestamp (PR #28414 fixed a cross-series collision bug here)
     - RRULE expansion with `MAX_SAFE_ITERATIONS` cap to prevent CPU DoS
     - DST handling via dayjs.tz instead of static VTIMEZONE offset
     - Never log raw ICS URLs since they contain the CacheKey secret

4. **Setup page** - Proton-branded with instructions:
   "Open Proton Calendar > Settings > Share via link > Full view > Copy link >
   Paste here." Note that paid Proton plan is required.

5. **Generated file updates** - `apps.metadata.generated.ts`,
   `apps.server.generated.ts`, `calendar.services.generated.ts`,
   `bookerApps.metadata.generated.ts`

## Lessons from prior PRs

I've gone through every security flag across all 6 PRs and my plan addresses
all of them:
- ICS URL leaking into logs (P1, flagged on 4 PRs)
- Domain allowlist bypass via endsWith (P1, #28372)
- Missing auth guard on add.ts (P1, #27863, #28414)
- SSRF via HTTP redirects (P1, #28414)
- RRULE expansion DoS (P1, #28414)
- DST/timezone bugs (P2, multiple PRs)
- Cancelled-event ghost blocking (functional bug, multiple PRs)

## What you'll get

- PR with screenshots showing Proton busy slots blocking Cal.com availability
- CLA signed
- Clean code, no unnecessary comments
- Tests for CalendarService (ghost event filtering, RRULE cap, domain
  validation, URL redaction)

ETA: ~5 days.

**Comment by tribhuwan-kumar** [comment]
@romitg2 

is this still open or is it already completed??

**Comment by romitg2** [comment]
open for contributions.

> [@romitg2](https://github.com/romitg2)
> 
> is this still open or is it already completed??

**Comment by atukumbwike-cmd** [comment]
attempt #5756

**Comment by atukumbwike-cmd** [comment]
Implementation Plan for Proton Calendar Integration
​I would like to tackle this bounty by building a dedicated protoncalendar package within packages/app-store/.
​My approach:
​Wrapper Service: I will extend the existing IcsFeedCalendarService to create a Proton-specific service that handles the read-only ICS feeds provided by Proton.
​Branded UI: Implement a guided setup page in pages/setup/index.tsx specifically for Proton users, including clear instructions on how to obtain their ICS share link.
​Validation: Add domain validation to ensure the provided URLs are from proton.me or protonmail.com.
​Encryption: Ensure that the feed URLs are encrypted at rest using the existing credential management system.
​Clean Up: Filter out STATUS:CANCELLED events which are common in Proton ICS exports to ensure accurate availability.
​I'm ready to start and provide a high-quality PR with a demo video.

**Comment by mahenoorsalat** [comment]
/attempt #5756

I went through the full issue thread, previous PRs, and Proton’s current limitations before proposing an approach.

Understanding the constraint:
Proton Calendar does not support write access (no public API / CalDAV), and ICS links are read-only. So a full 2-way sync like Google/Outlook is not feasible. The integration should focus on availability (free/busy) rather than event creation.

Proposed approach:
Instead of building everything from scratch, I’ll create a dedicated protoncalendar app by wrapping the existing ICS feed infrastructure and handling Proton-specific edge cases.

Implementation plan:

Create packages/app-store/protoncalendar/
Extend existing ICSFeedCalendarService for reuse (no duplication)
Implement Proton-specific handling:
Filter STATUS:CANCELLED events (ghost blocking issue)
Handle recurring events + EXDATE properly
Strict domain validation (calendar.proton.me only)
Ensure ICS URLs are encrypted at rest (since they act like access tokens)
Setup UI:
Guided onboarding (how to generate Proton ICS link)
Clear messaging: read-only + sync delay limitations
Default “skip writing events” behavior
Calendar behavior:
Availability works fully
Event creation becomes a no-op with fallback to .ics email attachment
Security + reliability:
Prevent SSRF via controlled fetch
Avoid logging sensitive ICS URLs
Add basic tests for parsing + filtering logic

Focus:

Minimal, clean diff (aligned with existing app-store patterns)
No unnecessary abstractions
Production-ready + easy to review

I’ll also include a short demo video showing:

Setup flow
Availability blocking using Proton data
Booking flow with fallback handling

If there’s anything specific you’d like prioritized (UX vs strict backend correctness), let me know before I start implementation.

**Comment by lky141319** [comment]
Hi! 👋 I can help with this.

I'm a developer available for hire immediately. I can handle this task efficiently.

#### Why me?
- Available to start right now
- Fast delivery
- Flexible pricing (USDT accepted)

Let me know if we can work together! 💰

**Comment by MilosM348** [comment]
/attempt #5756

Hi @alishaz-polymath — picking up the dedicated Proton Calendar app you and @PeerRich described back when this was reopened, and that @Rhan2020 attempted in #27863 (closed). I'd like to ship it with the two `cubic-dev-ai` review issues from that PR fixed plus test coverage. Plan below — happy to adjust before opening the PR.

### Scope

A new `packages/app-store/protoncalendar/` app forked from the generic ICS Feed app, so Proton-specific quirks don't leak back into the shared ICS code path used by other providers.

Read-only by design — Proton has confirmed they will not expose a third-party write API ([thread](https://github.com/calcom/cal.diy/issues/5756#issuecomment-1600571667)). The "Limited view (free/busy)" ICS feed is the API. Confirmation of attendees stays on email, which already works today.

### Three Proton fixes the generic ICS app does not handle

1. **Ghost events** — `STATUS:CANCELLED` VEVENTs are filtered out. (The behavior @colans has been reporting since 2024.)
2. **Cancelled recurring instances** — `RECURRENCE-ID + STATUS:CANCELLED` is honored as an EXDATE so cancelled occurrences of a master RRULE no longer count as busy.
3. **DST correctness** — each occurrence's wall-clock is resolved through `@calcom/dayjs.tz(iso, ianaTzId)` instead of building a single static STANDARD `VTIMEZONE` block from one event's offset (the cubic-dev-ai P2 finding on #27863). Bookings around a DST transition will now sit on the right side of the boundary.

### Security

- `api/add.ts` enforces a session check before touching Prisma. The reference `ics-feedcalendar/api/add.ts` calls `prisma.user.findFirstOrThrow({ where: { id: req.session?.user?.id } })`, and Prisma silently drops `where` clauses where the value is `undefined`, so an unauthenticated POST resolves to the first user in the table. New handler returns 401 immediately if no session. (cubic-dev-ai P1 finding on #27863.)
- URL host is validated against `calendar.proton.me` / `calendar.protonmail.com` / `calendar.pm.me`. Non-Proton URLs route the user back to the generic ICS Feed app.

### Tests

- `lib/__tests__/isProtonCalendarUrl.test.ts` — host whitelist, look-alike rejection, http vs https, case-insensitive host comparison.
- Will add `CalendarService.test.ts` covering the ghost-event filter, recurring-exception handling, and DST-boundary occurrences with vitest fixtures of real Proton ICS payloads.

ETA ~5 days from a thumbs-up. Happy to adjust scope (drop the Proton-specific app and patch the existing ICS Feed app instead, if you'd prefer that direction).

**Comment by ishaannk** [comment]
I’m exploring the current calendar integration architecture and reviewing the existing CalDAV/ICS handling flow. My plan is to prototype Proton-compatible availability handling through ICS-based synchronization and evaluate the minimal changes needed for booking confirmation support before opening a PR.
/attempt #5756

**Comment by dheerajsingh0773-ctrl** [comment]
"Hi! I'm a Web Developer and I'm very interested in this $200 bounty (#3857). I've checked the requirements for supporting different k8s clusters/namespaces and I'm ready to work on it. Please assign this to me."
/attempt #3857
Isse Algora ka system tumhe register kar lega.

**Comment by dheerajsingh0773-ctrl** [comment]
/attempt #3857
(Isse Algora bot tumhara naam register kar lega ki tum is par kaam kar rahe ho).

**Comment by neilpotdukhe** [comment]
/attempt #5756

I am taking a scoped Proton Calendar implementation that reuses the existing read-only ICS feed infrastructure instead of waiting on a private Proton API.

Plan:
- Add a dedicated `packages/app-store/protoncalendar` calendar app with Proton metadata, API add handler, and CalendarService registration.
- Accept only Proton Calendar share URLs over HTTPS, store the feed URL encrypted like the existing ICS feed app, and validate the feed by listing calendars before creating the credential.
- Add a Proton-specific setup page with the read-only limitation and sync-delay guidance.
- Keep booking writes as read-only/no-op behavior via the existing ICS service path so availability works and booking creation does not fail because Proton cannot accept writes.
- Run focused type/lint checks for the changed package where possible and include `/claim #5756` in the PR body after verification.

**Comment by dheerajsingh0773-ctrl** [comment]
Hi Ioannis,
Thanks for the review. I will check the type errors locally right away, fix
them, and update the PR.
Best regards,
Dheeraj"

**Comment by quangnd2203** [comment]
Hi! I noticed there is already a linked PR/claim activity on this bounty. Is this still open for a new contributor, or is it effectively taken already? Thanks!

**Comment by neilpotdukhe** [comment]
Hi @quangnd2203, I have an active PR open for this now: #29350, with `/claim #5756` in the PR body.

It is currently pending maintainer review and the repo external-contributor CI gate (`run-ci`). I cannot speak for the maintainers on whether they want parallel attempts, but there is an active claimed implementation in review.

**Comment by ThaiTrevor** [comment]
C:/Program Files/Git/attempt #5756

**Comment by MyTH-zyxeon** [comment]
Maintainer-facing acceptance synthesis for #5756, without opening a duplicate implementation or adding a duplicate claim.

Current snapshot from the open Proton Calendar PR set:

- #29350 is open, mergeable, `mergeStateStatus=BLOCKED`, `reviewDecision=REVIEW_REQUIRED`, and includes `/claim #5756`. It appears to pursue the read-only Proton shared calendar / ICS feed route, with app setup, generated app-store registry updates, add API, calendar service, URL validation tests, encrypted credential handling, feed verification, and read-only booking behavior.
- #29449 is open, mergeable, `mergeStateStatus=BLOCKED`, `reviewDecision=REVIEW_REQUIRED`, and appears to be the newest broader CalDAV-oriented attempt. It includes app metadata, generated registry updates, add API work, calendar service tests, and seed/app-store wiring, but the CalDAV path should be explicitly checked for credential scope, secret handling, and whether write semantics are actually required for this bounty.
- #29419, #29445, #29345, and #29306 are also open and mergeable, but still blocked/review-required. They overlap substantially on app setup, Proton calendar API/service code, validation, generated app-store files, and registration/i18n details.

Suggested maintainer decision point:

- If #5756 only requires showing Proton availability in Cal.com, the lower-risk acceptance path is a read-only shared calendar / ICS feed implementation, with create/update/delete event operations intentionally no-op or unsupported.
- If #5756 requires full authenticated CalDAV write support, please say that explicitly so competing attempts can be judged against credential storage, refresh/auth behavior, write semantics, and security expectations instead of just feed import behavior.

Acceptance checklist before selecting a winner:

- Pick one public app slug/route convention (`protoncalendar` versus `proton-calendar`) and reject duplicate registry variants.
- Trigger the required maintainer CI gate (`run-ci`) on the preferred PR; the visible open candidates are still blocked/review-required from the external contributor path.
- Verify encrypted storage of the Proton URL/credential material and confirm no raw shared calendar URL is logged.
- Smoke-test valid and invalid Proton shared-calendar URLs through the add API.
- Check recurring/cancelled events, redirect/timeout handling, and malformed ICS responses.
- Confirm booking create/update/delete semantics match the chosen scope: no-op/unsupported for read-only ICS, or real write behavior for authenticated CalDAV.
- Verify generated app-store registry and i18n changes are consistent and not stale.
- Avoid duplicate payout confusion by selecting one claim/PR once the acceptance scope is clear.

I am not claiming #5756 here and did not open a competing PR. This is only an acceptance/triage synthesis to help unblock a maintainer decision across the crowded open PR set.

**Comment by tedashikode** [comment]
@xdaluca Please, I want to solve this issue. Is it still open?

**Comment by Shaidyk** [comment]
/claim #5756

PR: https://github.com/calcom/cal.diy/pull/29582

**Comment by GAURAV07C** [comment]
@bandhan-majumder   is this still open or is it already completed?

**Comment by bandhan-majumder** [comment]
as this will require a lot of back and forth communication between contributor and team members, closing this for now.

more: https://cal.com/blog/cal-com-goes-closed-source-why
